File: Awesome.AI.Source/Awesome.AI/CoreSystems/Environment.cs
using Awesome.AI.Common;
using Awesome.AI.Core;
using Awesome.AI.Core.Spaces;
using Awesome.AI.CoreInternals;
using Awesome.AI.Variables;
using static Awesome.AI.Variables.Enums;

namespace Awesome.AI.CoreSystems
{
    class SimpleAgent
    {
        /*
         * family, friends and just persons to interact with
         * */

        private bool Property { get; set; }

        private TheMind mind;
        private SimpleAgent() { }
        public SimpleAgent(TheMind mind)
        {
            this.mind = mind;
        }

        public double SimulateDirection()
        {
            //bool samp = CONST.SAMPLE20.RandomSample(mind);

            double rand = mind.rand.MyRandomDouble(1)[0];

            double down = rand <= 0.5d ? -1.0d : 1.0d;

            return down;
        }

        public bool SimulateYes()
        {
            //bool samp = CONST.SAMPLE20.RandomSample(mind);

            double dir = SimulateDirection();

            bool down = dir <= 0.0d;

            return down;
        }

        public void SetProperty(bool _b)
        {
            Property = _b;
        }
    }

    class TimeLine
    {
        /*
         * youth, adolecence, events, learning..
         * for creating memories and setting up the system 
         * */
    }


    /*
     * rooms at home, work and other
     * ie. at work you mostly think about work and so on
     * */
    public class Occupasion
    {
        public string name { get; set; }
        public int max_epochs { get; set; }
        public List<string> values { get; set; }
    }

    public class Ticket
    {
        public string t_name { get; set; }

        public Ticket(string _n)
        {
            t_name = _n;
        }
    }

    public class MyInternal// aka MapMind
    {
        private TheMind mind;
        private MyInternal() { }
        public MyInternal(TheMind mind)
        {
            this.mind = mind;
        }

        public List<Occupasion> occus = new List<Occupasion>();//this is the map

        private Occupasion occu = new Occupasion() { name = "init", max_epochs = 10, values = null };
        private bool run = false;
        private int epoch_old = -1;
        public int epoch_count = 0;
        public int epoch_stop = -1;


        public Occupasion Occu
        {
            get
            {
                /*
                 * run is only true once per cycle
                 * */

                run = mind.epochs != epoch_old;
                epoch_old = mind.epochs;

                if (run)
                {
                    switch (mind.parms_current.occupasion)
                    {
                        case OCCUPASION.FIXED:
                            Lookup lookup = new Lookup();
                            MINDS mindtype = mind.mindtype;
                            List<string> list = lookup.GetOCCU(mindtype, 0, out string _o);
                            occu = new Occupasion() { name = _o, max_epochs = 30, values = list };
                            break;
                        case OCCUPASION.DYNAMIC:

                            /*
                             * rand should be set according to hobbys, mood, location, interests etc..
                             * ..maybe not
                             * */

                            if (epoch_count <= epoch_stop)
                                break;

                            epoch_count = 0;
                            epoch_stop = mind.rand.MyRandomInt(1, occu.max_epochs)[0];
                            int index = mind.rand.MyRandomInt(1, occus.Count - 1)[0];

                            occu = occus[index];

                            break;
                        default:
                            throw new Exception("Occu");
                    }

                    epoch_count++;
                }

                return occu;
            }
        }

        public bool Valid(UNIT _u)
        {
            if (mind.z_current == "z_noise")
                return true;

            if (!Filters.FilterUnit(mind, FILTERUNIT.NONE, FILTERTYPE.THREE, _u))
                return true;

            try
            {
                string name = Occu.name;
                string sub = mind.hub.GetSubject(_u) ?? "";

                if (name == "")
                    return false;

                if (sub == "")
                    return false;

                Occupasion occu = occus.Where(x => x.name == name).First();
                List<string> _hubs = occu.values;
                bool res = _hubs.Contains(sub);

                return res;
            }
            catch
            {
                return false;
            }
        }

        //process occupation
        public void Setup()
        {
            /*
             * these should be set according to hobbys, mood, location, interests etc..
             * */

            occus = new List<Occupasion>();

            Lookup lookup = new Lookup();
            MINDS[] minds = { MINDS.ROBERTA, MINDS.ANDREW };
            int occu_count = lookup.occupasions.Length;
            
            foreach (MINDS mind in minds)
            {
                for (int i = 0; i < occu_count; i++)
                {
                    List<string> list = lookup.GetOCCU(mind, i, out string occu);
                    occus.Add(new Occupasion() { name = occu, max_epochs = 30, values = list });                
                }
            }
        }

        public void Reset()
        {
            if (mind.z_current == "z_noise")
                return;

            if (mind.STATE == STATE.QUICKDECISION)
                return;

            if (mind.parms_current.validation != VALIDATION.EXTERNAL)
                Setup();
        }
    }


    public class MyExternal// aka MapWorld
    {
        public class Tag
        {
            public string t_name { get; set; }

            public Tag(string _n)
            {
                t_name = _n;
            }
        }

        public List<Tag> tags = new List<Tag>();//this is the map

        private TheMind mind;
        private MyExternal() { }
        public MyExternal(TheMind mind)
        {
            this.mind = mind;
        }

        public bool Valid(UNIT _u)
        {
            if (mind.z_current == "z_noise")
                return true;

            if (_u.ticket.IsNull())
                throw new Exception("Valid");

            if (_u.IsDECISION())
                return true;

            if (_u.IsQUICKDECISION())
                return true;

            tags = tags.Where(x => x != null).ToList();

            double scale = 0.0d;

            //bool t_name = _u.ticket.t_name == "SPECIAL";
            bool tags_hit = tags.Where(x => x.t_name == _u.ticket.t_name).Any();

            bool hit = /*t_name || */tags_hit;// || focus;

            return hit;
        }

        //setup input
        private void Setup(MINDS mindtype, bool onlyeven)
        {
            tags = new List<Tag>();
            tags.Add(new Tag("SPECIAL"));

            List<string> list;

            list = mind.memory.Tags(mindtype);

            foreach (string s in list)
            {
                for (int i = 1; i <= CONST.NUMBER_OF_UNITS; i++)
                {
                    if (onlyeven && i % 2 == 0)
                        continue;

                    tags.Add(new Tag("" + s + i));
                }
            }
        }

        public void Reset()
        {
            if (mind.z_current == "z_noise")
                return;

            if (mind.parms_current.validation != VALIDATION.INTERNAL)
            {
                //mind.stats.Reset();

                tags = new List<Tag>();
                switch (mind.parms_current.tags)
                {
                    case TAGS.ALL: Setup(mind.mindtype, false); break;
                    case TAGS.EVEN: Setup(mind.mindtype, true); break;
                    default: throw new Exception("Reset");
                }
            }
        }
    }
}

File: Awesome.AI.Python/CoreSystems/Environment.py
"""Internal/external environment maps ported from ``Environment.cs``."""
from dataclasses import dataclass, field

from awesome_ai.Core.Internals.Lookup import Lookup
from awesome_ai.Variables.Constants import CONST
from awesome_ai.Variables.Enums import MINDS, OCCUPASION, STATE, TAGS, VALIDATION


class SimpleAgent:
    def __init__(self, mind): self.mind, self.Property = mind, False
    def SimulateDeltaVelocity(self): return self.mind.rand.MyRandomDouble(1)[0]
    def SimulateDirection(self): return -1.0 if self.mind.rand.MyRandomDouble(1)[0] <= 0.5 else 1.0
    def SimulateYes(self): return self.SimulateDirection() <= 0.0
    def SetProperty(self, _b): self.Property = _b


# Backward-compatible name used by the first scaffold.
Environment = SimpleAgent


class TimeLine:
    """Reserved C# timeline model for later memory/event expansion."""


@dataclass
class Occupasion:
    name: str = ""
    max_epochs: int = 0
    values: list[str] = field(default_factory=list)


@dataclass
class Ticket:
    t_name: str = ""


class MyInternal:
    """MapMind: selects the current internal occupation and its valid hubs."""

    def __init__(self, mind=None):
        self.mind = mind
        self.occus = []
        self.occu = Occupasion(name="init", max_epochs=10, values=[])
        self.run = False
        self.epoch_old = -1
        self.epoch_count = 0
        self.epoch_stop = -1

    @property
    def Occu(self):
        self.run = self.mind.epochs != self.epoch_old
        self.epoch_old = self.mind.epochs
        if not self.run:
            return self.occu

        if self.mind.bot.occupasion == OCCUPASION.FIXED:
            hubs, occupation = Lookup().GetHUBS(self.mind.mindtype, 0, with_occupation=True)
            self.occu = Occupasion(name=occupation, max_epochs=30, values=hubs)
        elif self.mind.bot.occupasion == OCCUPASION.DYNAMIC:
            if self.epoch_count > self.epoch_stop:
                self.epoch_count = 0
                maximum = max(1, self.occu.max_epochs)
                self.epoch_stop = self.mind.rand.Next(1, maximum + 1)
                if self.occus:
                    self.occu = self.occus[self.mind.rand.Next(len(self.occus))]
        else:
            raise RuntimeError("Occu")

        self.epoch_count += 1
        return self.occu

    def Valid(self, unit):
        if unit is None:
            return False
        if unit.IsDECISION():
            return True
        try:
            occupation = self.Occu
            subject = self.mind.hub.GetSubject(unit) or ""
            if not occupation.name or not subject:
                return False
            configured = next(item for item in self.occus if item.name == occupation.name)
            return subject in configured.values
        except (StopIteration, AttributeError, TypeError):
            return False

    def Setup(self):
        self.occus = []
        lookup = Lookup()
        mindtypes = (MINDS.BASIC,) if self.mind.mindtype == MINDS.BASIC else (MINDS.ROBERTA, MINDS.ANDREW)
        for mindtype in mindtypes:
            for index in range(len(CONST.occupasions)):
                hubs, occupation = lookup.GetHUBS(mindtype, index, with_occupation=True)
                self.occus.append(Occupasion(name=occupation, max_epochs=30, values=hubs))

    def Reset(self):
        if self.mind.STATE != STATE.QUICKDECISION and self.mind.bot.validation != VALIDATION.EXTERNAL:
            self.Setup()
            # HubSpace consults Occu during Python construction; restart the
            # epoch gate so the populated map is selected on the next access.
            self.epoch_old = -1
            self.epoch_count = 0
            self.epoch_stop = -1


class MyExternal:
    @dataclass
    class Tag:
        t_name: str = ""
    def __init__(self, mind=None):
        self.mind = mind
        self.tags = []

    @staticmethod
    def _ticket_name(unit):
        ticket = unit.ticket
        return ticket.t_name if hasattr(ticket, "t_name") else str(ticket or "")

    def Valid(self, unit):
        if unit is None:
            return False
        if unit.IsDECISION() or unit.IsQDECISION():
            return True
        ticket_name = self._ticket_name(unit)
        if not ticket_name:
            raise RuntimeError("Valid")
        self.tags = [tag for tag in self.tags if tag is not None]
        return any(tag.t_name == ticket_name for tag in self.tags)

    def Setup(self, mindtype, onlyeven):
        self.tags = [self.Tag("SPECIAL")]
        for subject in self.mind.memory.Tags(mindtype):
            for number in range(1, CONST.NUMBER_OF_UNITS + 1):
                if onlyeven and number % 2 == 0:
                    continue
                self.tags.append(self.Tag(f"{subject}{number}"))

    def Reset(self):
        if self.mind.bot.validation == VALIDATION.INTERNAL:
            return
        if self.mind.bot.tags == TAGS.ALL:
            self.Setup(self.mind.mindtype, False)
        elif self.mind.bot.tags == TAGS.EVEN:
            self.Setup(self.mind.mindtype, True)
        else:
            raise RuntimeError("Reset")
